Cannot watch build resources in Openshift 3.6

Related to [this issue](https://github.com/fabric8io/fabric8-maven-plugin/issues/1016) in the fabric8-maven-plugin.

The watch is always done with a resource version number that is fetched from the build list (not from the actual build) on the first attempt. Everything is looked up from the new API endpoint `apis/build.openshift.io`.

The error is like:
```
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Timeout: Too large resource version: 19393279, current: 19393278
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$2.onFailure(WatchConnectionManager.java:191)
	at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:543)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:185)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:141)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

From my tests, it seems that the resource version number in the list object **is independent** from the resource version number of the actual build object (in the new API).

So the first revision number should not be taken from the list, but from the actual build. I don't know why it is taken from the list.

Using a empty revision number should be a viable option on the first attempt to create the watch. While on a subsequent reconnection, the client should use the last revision number received from the server.

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java
@@ -130,11 +130,11 @@ public class WatchConnectionManager<T extends HasMetadata, L extends KubernetesR
       }
     }
 
-    httpUrlBuilder.addQueryParameter("watch", "true");
-
     if (this.resourceVersion.get() != null) {
       httpUrlBuilder.addQueryParameter("resourceVersion", this.resourceVersion.get());
     }
+
+    httpUrlBuilder.addQueryParameter("watch", "true");
 
     Request request = new Request.Builder()
       .get()

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchConnectionManager.java
@@ -82,12 +82,7 @@ public class WatchConnectionManager<T extends HasMetadata, L extends KubernetesR
   private OkHttpClient clonedClient;
 
   public WatchConnectionManager(final OkHttpClient client, final BaseOperation<T, L, ?, ?> baseOperation, final String version, final Watcher<T> watcher, final int reconnectInterval, final int reconnectLimit, long websocketTimeout) throws MalformedURLException {
-    if (version == null) {
-      L currentList = baseOperation.list();
-      this.resourceVersion = new AtomicReference<>(currentList.getMetadata().getResourceVersion());
-    } else {
-      this.resourceVersion = new AtomicReference<>(version);
-    }
+    this.resourceVersion = new AtomicReference<>(version); // may be a reference to null
     this.baseOperation = baseOperation;
     this.watcher = watcher;
     this.reconnectInterval = reconnectInterval;
@@ -135,9 +130,11 @@ public class WatchConnectionManager<T extends HasMetadata, L extends KubernetesR
       }
     }
 
-    httpUrlBuilder
-      .addQueryParameter("resourceVersion", this.resourceVersion.get())
-      .addQueryParameter("watch", "true");
+    httpUrlBuilder.addQueryParameter("watch", "true");
+
+    if (this.resourceVersion.get() != null) {
+      httpUrlBuilder.addQueryParameter("resourceVersion", this.resourceVersion.get());
+    }
 
     Request request = new Request.Builder()
       .get()
@@ -223,7 +220,7 @@ public class WatchConnectionManager<T extends HasMetadata, L extends KubernetesR
             // Dirty cast - should always be valid though
             String currentResourceVersion = resourceVersion.get();
             String newResourceVersion = ((HasMetadata) obj).getMetadata().getResourceVersion();
-            if (currentResourceVersion.compareTo(newResourceVersion) < 0) {
+            if (currentResourceVersion == null || currentResourceVersion.compareTo(newResourceVersion) < 0) {
               resourceVersion.compareAndSet(currentResourceVersion, newResourceVersion);
             }
             Watcher.Action action = Watcher.Action.valueOf(event.getType());
@@ -235,7 +232,7 @@ public class WatchConnectionManager<T extends HasMetadata, L extends KubernetesR
             // Dirty cast - should always be valid though
             String currentResourceVersion = resourceVersion.get();
             String newResourceVersion = list.getMetadata().getResourceVersion();
-            if (currentResourceVersion.compareTo(newResourceVersion) < 0) {
+            if (currentResourceVersion == null || currentResourceVersion.compareTo(newResourceVersion) < 0) {
               resourceVersion.compareAndSet(currentResourceVersion, newResourceVersion);
             }
             Watcher.Action action = Watcher.Action.valueOf(event.getType());

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java
@@ -78,12 +78,7 @@ public class WatchHTTPManager<T extends HasMetadata, L extends KubernetesResourc
                           final int reconnectLimit, long connectTimeout)
     throws MalformedURLException {
 
-    if (version == null) {
-      L currentList = baseOperation.list();
-      this.resourceVersion = new AtomicReference<>(currentList.getMetadata().getResourceVersion());
-    } else {
-      this.resourceVersion = new AtomicReference<>(version);
-    }
+    this.resourceVersion = new AtomicReference<>(version); // may be a reference to null
     this.baseOperation = baseOperation;
     this.watcher = watcher;
     this.reconnectInterval = reconnectInterval;
@@ -138,9 +133,11 @@ public class WatchHTTPManager<T extends HasMetadata, L extends KubernetesResourc
       httpUrlBuilder.addQueryParameter("fieldSelector", fieldQueryString);
     }
 
-    httpUrlBuilder
-      .addQueryParameter("resourceVersion", this.resourceVersion.get())
-      .addQueryParameter("watch", "true");
+    httpUrlBuilder.addQueryParameter("watch", "true");
+
+    if (this.resourceVersion.get() != null) {
+      httpUrlBuilder.addQueryParameter("resourceVersion", this.resourceVersion.get());
+    }
 
     final Request request = new Request.Builder()
       .get()
@@ -239,7 +236,7 @@ public class WatchHTTPManager<T extends HasMetadata, L extends KubernetesResourc
         // Dirty cast - should always be valid though
         String currentResourceVersion = resourceVersion.get();
         String newResourceVersion = ((HasMetadata) obj).getMetadata().getResourceVersion();
-        if (currentResourceVersion.compareTo(newResourceVersion) < 0) {
+        if (currentResourceVersion == null || currentResourceVersion.compareTo(newResourceVersion) < 0) {
           resourceVersion.compareAndSet(currentResourceVersion, newResourceVersion);
         }
         Watcher.Action action = Watcher.Action.valueOf(event.getType());
@@ -250,7 +247,7 @@ public class WatchHTTPManager<T extends HasMetadata, L extends KubernetesResourc
           // Dirty cast - should always be valid though
           String currentResourceVersion = resourceVersion.get();
           String newResourceVersion = list.getMetadata().getResourceVersion();
-          if (currentResourceVersion.compareTo(newResourceVersion) < 0) {
+          if (currentResourceVersion == null || currentResourceVersion.compareTo(newResourceVersion) < 0) {
             resourceVersion.compareAndSet(currentResourceVersion, newResourceVersion);
           }
           Watcher.Action action = Watcher.Action.valueOf(event.getType());

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/WatchHTTPManager.java
@@ -133,11 +133,11 @@ public class WatchHTTPManager<T extends HasMetadata, L extends KubernetesResourc
       httpUrlBuilder.addQueryParameter("fieldSelector", fieldQueryString);
     }
 
-    httpUrlBuilder.addQueryParameter("watch", "true");
-
     if (this.resourceVersion.get() != null) {
       httpUrlBuilder.addQueryParameter("resourceVersion", this.resourceVersion.get());
     }
+
+    httpUrlBuilder.addQueryParameter("watch", "true");
 
     final Request request = new Request.Builder()
       .get()

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/PodTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/PodTest.java
@@ -268,7 +268,7 @@ public class PodTest {
             .build()
     ).once();
 
-   server.expect().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true")
+   server.expect().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true")
             .andUpgradeToWebSocket()
             .open()
             .waitFor(15000).andEmit(new WatchEvent(pod1, "DELETED"))
@@ -350,7 +350,7 @@ public class PodTest {
       .endMetadata()
       .withItems(notReady).build()).once();
 
-    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true").andUpgradeToWebSocket()
+    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true").andUpgradeToWebSocket()
       .open()
       .waitFor(1000).andEmit(new WatchEvent(ready, "MODIFIED"))
       .done()

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/ResourceTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/ResourceTest.java
@@ -98,7 +98,7 @@ public class ResourceTest {
       server.expect().get().withPath("/api/v1/namespaces/test/pods").andReturn(200, pod1).once();
       server.expect().post().withPath("/api/v1/namespaces/test/pods").andReturn(201, pod1).once();
 
-     server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true").andUpgradeToWebSocket()
+     server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true").andUpgradeToWebSocket()
         .open()
           .waitFor(1000).andEmit(new WatchEvent(pod1, "DELETED"))
         .done()
@@ -149,7 +149,7 @@ public class ResourceTest {
 
     server.expect().get().withPath("/api/v1/namespaces/test/pods").andReturn(200, noReady).once();
 
-    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true").andUpgradeToWebSocket()
+    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true").andUpgradeToWebSocket()
       .open()
       .waitFor(1000).andEmit(new WatchEvent(ready, "MODIFIED"))
       .done()
@@ -187,7 +187,7 @@ public class ResourceTest {
     server.expect().get().withPath("/api/v1/namespaces/test/pods").andReturn(200, noReady).once();
     server.expect().post().withPath("/api/v1/namespaces/test/pods").andReturn(201, noReady).once();
 
-    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true").andUpgradeToWebSocket()
+    server.expect().get().withPath("/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true").andUpgradeToWebSocket()
       .open()
       .waitFor(1000).andEmit(new WatchEvent(ready, "MODIFIED"))
       .done()

